packages/fast-components/src/form-associated/index.ts

```diff
@@ -0,0 +1,263 @@
+     * form is submitted
+     */
+    public get willValidate(): boolean {
+        return FormAssociated.formAssociated
+            ? this.elementInternals.willValidate
+            : this.proxy.willValidate;
+    }
+
+    /**
+     * A reference to all associated label elements
+     */
+    public get labels(): Node[] {
+        if (FormAssociated.formAssociated) {
+            return Array.from(this.elementInternals.labels);
+        } else if (
+            this.proxy instanceof HTMLElement &&
+            this.proxy.ownerDocument &&
+            this.id
+        ) {
+            // Labels associated by wraping the element: <label><custom-element></custom-element></label>
+            const parentLabels = this.proxy.labels;
+            // Labels associated using the `for` attribute
+            const forLabels = Array.from(
+                this.proxy.ownerDocument.querySelectorAll(`[for='${this.id}']`)
+            );
+
+            return !!parentLabels
+                ? forLabels.concat(Array.from(parentLabels))
+                : forLabels;
+        } else {
+            return [];
+        }
+    }
+
+    /**
+     * The value of the element to be associated with the form
+     */
+    @attr
+    public value: string = "";
+
+    /**
+     * Whether the element should be autofocused
+     */
+    @attr
+    public autofocus: boolean; // Map to proxy element
+
+    @attr
+    public disabled: boolean = false;
+    private disabledChanged(): void {
+        if (this.proxy instanceof HTMLElement) {
+            this.proxy.disabled = this.disabled === ("" as any); // TODO: https://github.com/microsoft/fast-dna/issues/2742
+        }
+    }
+
+    @attr
+    public name: string;
+    private nameChanged(): void {
+        if (this.proxy instanceof HTMLElement) {
+            this.proxy.name = this.name;
+        }
+    }
+
+    /**
+     * Require the field prior to form submission
+     */
+    @attr
+    public required: boolean = false;
+    private requiredChanged(): void {
+        if (this.proxy instanceof HTMLElement) {
+            this.proxy.required = this.required === ("" as any); // TODO: https://github.com/microsoft/fast-dna/issues/2742
+        }
+    }
+
+    /**
+     * The proxy element provided by
+     */
+    protected abstract proxy: T;
+
+    /**
+     * The element internals object. Will only exist
+     * in browsers supporting the attachInternals API
+     */
+    protected elementInternals: ElementInternals;
+
+    /**
+     * These are events that are still fired by the proxy
+     * element based on user / programmatic interaction.
+     *
+     * The proxy implementation should be transparent to
+     * the app author, so block these events from emitting.
+     */
+    private proxyEventsToBlock = ["change", "click"];
+
+    constructor() {
+        super();
+
+        if (FormAssociated.formAssociated) {
+            this.elementInternals = (this as any).attachInternals();

```

Comment:
As a follow-up PR, I wonder if we could create an API like `controller.attachInternals(this)` which would attach and return the internals or create the proxy and a fake internals implementation that has many of the properties above on it.

Reply:
I think that is likely the best long-term solution since the ElementInternals object is likely to expand over time.

Reply:
Ah. If the proxy isn't created until after they are set...yeah that makes sense.